### PR TITLE
Adds audit for fixed media sizes

### PR DESCRIPTION
**Pull request by paullewis**
This looks at video, image, and SVG elements and ensures that they have width and height dimensions specified in their attributes or via the CSS rules that match to them.

### CODE REVIEW

**Comment by paullewis** [comment]
Same as #509 for flakiness

File: lighthouse-core/driver/gatherers/media-sized.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      }
+    }
+
+    var definesSizes = rules.find(r => r.style.width !== '' && r.style.height !== '');
+    return (definesSizes !== undefined);
+  };
+
+  var unsizedElements =
+      [...document.querySelectorAll('img, video, svg')]
+      .filter(el => !hasSizeAttributes(el))
```
Comment (paulirish) [inline lighthouse-core/driver/gatherers/media-sized.js]: A number of our gatherers report a bit more verbosely, and let the audits do the conditionals. I'd be into reporting an object for each img/video/svg element with width/height properties. And then leave it to the audit to evaluate if the sizes are defined.

That would save us some work when doing an audit for checking media elements for autoplay, or something.

But it's not a big deal for me, and we can refactor this bit later as well.

**Comment by paulirish** [comment]
BTW here's the equivalent rule in devtools: https://github.com/ChromeDevTools/devtools-frontend/blob/master/front_end/audits/AuditRules.js#L842-L983

File: lighthouse-core/driver/gatherers/media-sized.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      .map(el => {
+        let fileName;
+        if (el.src) {
```
Comment (brendankenny) [inline lighthouse-core/driver/gatherers/media-sized.js]: if it doesn't have a `src`, do we care about it?

**Comment by paulirish** [comment]
Areas of potential improvement:

* elements that are pos absolute (or its offsetParent (chain)) is pos absolute, we could discard it as it wont affect layout.
* the devtools impl does checks width and height separately in case they are set in different style rules. not a bad idea
* can we support the responsive images with fixed aspect ratio technique?
  * e.g. http://www.mademyday.de/css-height-equals-width-with-pure-css.html http://daverupert.com/2015/12/intrinsic-placeholders-with-picture/ https://justmarkup.com/log/2015/11/definining-aspect-ratio-to-prevent-reflow/
  * there's the padding-bottom hack and then a few others. 
* we might want to use https://chromedevtools.github.io/debugger-protocol-viewer/tot/CSS/#method-getMatchedStylesForNode instead of doc.stylesheets because I think doc.stylesheets excludes something.... 

None of these are blockers IMO, but it'd make this more robust.

How do others feel? Land this and improve later?

File: lighthouse-core/closure/typedefs/Artifacts.js [hunk 3]
```diff
@@ -79,3 +79,6 @@ Artifacts.prototype.ContentWidth;
 
 /** @type {!Array<string>} */
 Artifacts.prototype.CacheContents;
+
+/** @type {(!Array<!Object>|number)} */
```
Comment (brendankenny) [inline lighthouse-core/closure/typedefs/Artifacts.js]: `@type {(!Array<string>|number)}`?

File: lighthouse-core/driver/gatherers/media-sized.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+          fileName = new URL(el.src).pathname.split('/').slice(-1)[0];
+        }
+
+        return `${el.nodeName.toLowerCase()}${el.className ? '.' + el.className : ''}` +
```
Comment (brendankenny) [inline lighthouse-core/driver/gatherers/media-sized.js]: maybe build this over multiple steps? Hard to follow in this form

File: lighthouse-core/test/driver/gatherers/media-sized.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+const assert = require('assert');
+let mediaSizedGatherer;
+
+describe('Viewport gatherer', () => {
+  // Reset the Gatherer before each test.
+  beforeEach(() => {
+    mediaSizedGatherer = new MediaSizedGatherer();
+  });
+
+  it('returns an artifact', () => {
```
Comment (paulirish) [inline lighthouse-core/test/driver/gatherers/media-sized.js]: Do we want to test anything more substantial in here?

**Comment by samccone** [comment]
is this PR still active? if not perhaps we should close for now

**Comment by paulirish** [comment]
Let's close but can you open an issue to track this?

**Comment by samccone** [comment]
https://github.com/GoogleChrome/lighthouse/issues/802
